cpp/src/io/csv/csv_gpu.h

```diff
@@ -23,6 +23,99 @@ namespace cudf {
 namespace io {
 namespace csv {
 namespace gpu {
+
+/**
+ * CSV row parsing context
+ * NONE: No special context (normal parsing)
+ * QUOTE: Within a quoted field
+ * COMMENT: Within a comment line (discard every character until terminator)
+ * EOF: End state (EOF reached)
+ **/
+enum { ROW_CTX_NONE = 0, ROW_CTX_QUOTE = 1, ROW_CTX_COMMENT = 2, ROW_CTX_EOF = 3 };
+
```

Comment:
I'm finding it very hard to understand the gather implementation. Can you please include the description of the context format(s) in this header. As it is, it is very disjointed and hard to understand the hard-coded constants used in the implementation.

Reply:
Added a better description of the row context. It's basically just a 2-bit parsing state with row count (current row count and current parsing state), and the packed format just maps 4 possible input states to 4 possible output states (so the parsing is done for all possible input parsing states so it can be done in parallel without requiring knowledge of the initial input state, which is the part that causes serialization).

Reply:
[I added the description of each format above the corresponding `make_row_context` and `pack_row_contexts` helpers]